### ISSUE TITLE
Design decision to pass EINTR to user is flawed

### ISSUE THREAD

**Issue by franzhollerer** [zeromq/libzmq#3852]
I think the design decision to pass EINTR to the user is flawed and should be reconsidered for future versions. My arguments are similar to the arguments I found here:

- https://github.com/golang/go/issues/20400
- https://www.python.org/dev/peps/pep-0475

If libzmq fails to handle EINTR correctly and consistently within the library, how could it be expected from the user to do so?

On the one hand receiving a signal is a non-exceptional occurrence. Robust POSIX code must be prepared to handle EINTR. On the other hand EINTR is tightly coupled on the system call. A higher-level API may involve several system calls. Passing EINTR through various layer of abstraction presents the user with a serious problem. He cannot just restart the high level API function.

In my opinion EINTR must be handled locally where it occurs. The library should provide an behavior as when system calls are always restarted.

If the user needs some cleanup code before he exits the application, he should be advised not to rely on SIGINT and SIGTERM to interrupt system calls - as they are always restarted within the library (similar as when SA_RESTART would be set). Instead he should use timeouts, zmq_poll(), and make sure that he does not block in a system call forever.

To sum up: The user has no way to deal with EINTR appropriately, especially when high level APIs like czmq or cppmq are used. For robust code libzmq must deal with EINTR internally.

**Comment by bluca** [comment]
Without going into the details of whether this is right or wrong - such a change would not be backward compatible, so it cannot happen.

**Comment by franzhollerer** [comment]
Well, then I think the high level APIs should be aligned with the behavior of libzmq - which IMHO is currently not the case.

As far as I have seen czmq functions map EINTR to NULL, and in turn EINTR cannot be distinguished from other errors by the caller.

cppzmq handles EAGAIN internally, but throws an error_t for all other errors including EINTR. This pushes EINTR far away from the point where the problem occurred. I am quite sure that at the point where the exception is caught, the user has no glue how to recover from an EINTR.

**Comment by bluca** [comment]
Again without going into the merit of the proposal itself, those are not backward compatible changes either. Existing APIs cannot be changed.

**Comment by franzhollerer** [comment]
Thinking about backward compatibility in more details, I don't agree with the argument above.

The signature of the API functions will keep as is. EINTR will no longer be passed through different layers of abstraction - which is unmanageable for the application developer anyway.

The behavior slightly changes. But this could be controlled with a context option. If the option ZMQ_RESTART is set with zmq_ctx_set(), libzmq restarts interrupted system calls automatically, otherwise it passes EINTR to the user as ever.

This keeps the library backward compatible, but allows a less error-prone and more concise use in new applications.

**Comment by bluca** [comment]
A slight change in behavior is a change in behavior. Compatibility is about more than API definitions.

**Comment by franzhollerer** [comment]
Agreed. But I think I pointed out a way how backward compatibility could be achieved. ZMQ_RESTART would be just a new feature which can be turned on.
